### PR TITLE
COMP: Enable OpenSSL support in DCMTK build

### PR DESCRIPTION
**Pull request by jadh4v**
_Edited by @jcfr after [squashing commits](https://github.com/Slicer/Slicer/commits/c3b85a1728a79bf4a4d07c198c12098661312ad3/SuperBuild/External_DCMTK.cmake) originally associated with this pull request_

Add support for building DCMTK with OpenSSL support enabled introducing option `Slicer_USE_DCMTK_WITH_OPENSSL` set to `OFF` by default.

Update the DCMTK external project by introducing `EXTERNAL_PROJECT_OPTIONAL_CMAKE_ARGS`, similar to the `curl` project. This addition allows passing build configuration-specific arguments that are not determined at configuration time.

To ensure the OpenSSL libraries built within the Slicer SuperBuild are used, set the `DCMTK_USE_FIND_PACKAGE` option to `ON`. This enables customization of include directories and library paths through the following variables:
- Unix: `OPENSSL_INCLUDE_DIR`, `OPENSSL_SSL_LIBRARY`, `OPENSSL_CRYPTO_LIBRARY`
- Windows: `LIB_EAY_(RELEASE|DEBUG)`, `SSL_EAY_(RELEASE|DEBUG)`

<details><summary>Preview Description</summary>
<p>

# Description
This PR allows Slicer build system to enable SSL (OpenSSL) support in DCMTK. Enabling OpenSSL support in DCMTK requires setting `DCMTK_WITH_OPENSSL` to `ON` and `WITH_OPENSSLINC` to the installation path of OpenSSL library.
See: https://github.com/commontk/DCMTK/blob/patched-DCMTK-3.6.8_20241024/INSTALL#L584

# A few things that need fixing:
## 1. DCMTK expects [customized filenames](https://github.com/commontk/DCMTK/blob/patched-DCMTK-3.6.8_20241024/INSTALL#L626) for openssl library files
```
openssl:
       "dcmtkcrypto_d.lib" - debug version
       "dcmtkcrypto_o.lib" - release version (optimized)
       "dcmtkssl_d.lib"    - debug version
       "dcmtkssl_o.lib"    - release version (optimized)
```
We will need to either:

&nbsp;&nbsp;&nbsp; (a) create copies of these libraries in Slicer build tree with similar names, or
&nbsp;&nbsp;&nbsp; (b) explore the possibility of fixing the filenames in DCMTK's `3rdparty.cmake`

https://github.com/commontk/DCMTK/blob/patched-DCMTK-3.6.8_20241024/CMake/3rdparty.cmake#L305

## 2. Fix the `redefinition` build errors in DCMTK
UPDATE: FIXED by commit 53244ba4d0cde0ec40ef396904243b6a433bec96
```
26>D:\D\OpenSSL-install\Debug\include\openssl\ossl_typ.h(92,30): error C2371: 'EVP_MD_CTX': redefinition; different basic types [D:\D\DCMTK-build\dcmsign\libsrc\dcmdsig.vcxproj]
26>  (compiling source file '../../../DCMTK/dcmsign/libsrc/dcsignat.cc')
26>      D:\D\DCMTK\dcmsign\include\dcmtk\dcmsign\simdmac.h(39,30):
26>      see declaration of 'EVP_MD_CTX'
26>
26>  siautopr.cc
26>  simdmac.cc
26>D:\D\OpenSSL-install\Debug\include\openssl\ossl_typ.h(92,30): error C2371: 'EVP_MD_CTX': redefinition; different basic types [D:\D\DCMTK-build\dcmsign\libsrc\dcmdsig.vcxproj]
26>  (compiling source file '../../../DCMTK/dcmsign/libsrc/simdmac.cc')
26>      D:\D\DCMTK\dcmsign\include\dcmtk\dcmsign\simdmac.h(39,30):
26>      see declaration of 'EVP_MD_CTX'
26>
26>  sinullpr.cc
26>  sitstamp.cc
26>D:\D\OpenSSL-install\Debug\include\openssl\ossl_typ.h(92,30): error C2371: 'EVP_MD_CTX': redefinition; different basic types [D:\D\DCMTK-build\dcmsign\libsrc\dcmdsig.vcxproj]
26>  (compiling source file '../../../DCMTK/dcmsign/libsrc/sitstamp.cc')
26>      D:\D\DCMTK\dcmsign\include\dcmtk\dcmsign\simdmac.h(39,30):
26>      see declaration of 'EVP_MD_CTX'
```

</p>
</details>

### CODE REVIEW

File: SuperBuild/External_DCMTK.cmake [hunk 1]
```diff
@@ -65,7 +91,7 @@ if(NOT DEFINED DCMTK_DIR AND NOT Slicer_USE_SYSTEM_${proj})
       -DBUILD_SHARED_LIBS:BOOL=ON
       -DDCMTK_WITH_DOXYGEN:BOOL=OFF
       -DDCMTK_WITH_ZLIB:BOOL=OFF # see CTK github issue #25
-      -DDCMTK_WITH_OPENSSL:BOOL=OFF # see CTK github issue #25
+      -DDCMTK_WITH_OPENSSL:BOOL=${DCMTK_ENABLE_SSL} # see CTK github issue #25
```
Comment (jcfr) [inline SuperBuild/External_DCMTK.cmake:94]:
```suggestion
      -DDCMTK_WITH_OPENSSL:BOOL=${DCMTK_WITH_OPENSSL}
```

File: SuperBuild/External_DCMTK.cmake [hunk 2]
```diff
@@ -20,6 +27,25 @@ endif()
 if(NOT DEFINED DCMTK_DIR AND NOT Slicer_USE_SYSTEM_${proj})
   set(EXTERNAL_PROJECT_OPTIONAL_CMAKE_CACHE_ARGS)
 
+  if(DCMTK_ENABLE_SSL)
```
Comment (jcfr) [inline SuperBuild/External_DCMTK.cmake:30]:
```suggestion
  if(DCMTK_WITH_OPENSSL)
```

File: SuperBuild/External_DCMTK.cmake [hunk 3]
```diff
@@ -3,6 +3,13 @@ set(proj DCMTK)
 
 # Set dependency list
 set(${proj}_DEPENDENCIES "zlib")
+if(DCMTK_ENABLE_SSL)
```
Comment (jcfr) [inline SuperBuild/External_DCMTK.cmake:6]:
```suggestion
if(DCMTK_WITH_OPENSSL)
```

File: SuperBuild.cmake [hunk 4]
```diff
@@ -100,6 +100,7 @@ set(Slicer_DEPENDENCIES
   )
 
 set(CURL_ENABLE_SSL ${Slicer_USE_PYTHONQT_WITH_OPENSSL})
+set(DCMTK_ENABLE_SSL ${Slicer_USE_PYTHONQT_WITH_OPENSSL})
```
Comment (jcfr) [inline SuperBuild.cmake:103]:
```suggestion
set(DCMTK_WITH_OPENSSL ${Slicer_USE_PYTHONQT_WITH_OPENSSL})
```

**Comment by jadh4v** [comment]
@jcfr I have updated the PR. Please see the listed two problems in the description and suggest solutions for fixing them.

**Review by pieper — APPROVED** [review]
👍

File: SuperBuild/External_DCMTK.cmake [hunk 5]
```diff
@@ -3,6 +3,13 @@ set(proj DCMTK)
 
 # Set dependency list
 set(${proj}_DEPENDENCIES "zlib")
+if(DCMTK_WITH_OPENSSL)
+  if(NOT Slicer_USE_SYSTEM_${proj})
+    list(APPEND ${proj}_DEPENDENCIES OpenSSL)
+  else()
+    # XXX - Add a test checking if system DCMTK supports OpenSSL
```
Comment (lassoan) [inline SuperBuild/External_DCMTK.cmake:10]: We should add here something. At least log a warning message here.

**Review by lassoan — APPROVED** [review]
It looks good to me. Just added a small comment.

File: SuperBuild/External_DCMTK.cmake (hunk 5 above)
Comment (jcfr) [reply SuperBuild/External_DCMTK.cmake:10]: Thanks for the review :pray: 

I will address the following in a follow-up pull request:

```
$ ack "XXX - Add a test"
External_curl.cmake
10:    # XXX - Add a test checking if system curl support OpenSSL

External_DCMTK.cmake
10:    # XXX - Add a test checking if system DCMTK supports OpenSSL
```

**Pull request merged**